src/server/stream.rs

```diff
@@ -248,7 +251,23 @@ impl<S: SecuritySwitch> AsyncRead for SwitchingTlsStream<S> {}
 impl<S: SecuritySwitch> AsyncWrite for SwitchingTlsStream<S> {
     fn shutdown(&mut self) -> Poll<(), io::Error> {
         debug!("AsyncWrite shutdown <<{}>>", self.channel);
-        // TODO: Anything to do for the TLS stream??
+
+        let state = self.state.lock().unwrap().which_state(self.channel);
```

Comment:
Is it safe to unwrap here?

Reply:
the lock method will only give an error if the mutex was poisoned by a panic in another thread. In this case we probably want to panic too. Open for suggestions.

Reply:
I see. Just read [the Mutex documentation](https://doc.rust-lang.org/std/sync/struct.Mutex.html#poisoning). I guess we'll be fine unwrapping, then :+1: